src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java

```diff
@@ -22,101 +21,102 @@
   }
 
-  private List<InstructionCount> processActionType(final ActionType actionType) {
-    log.with("action_type", actionType.getName()).debug("Dealing with actionType");
-    final InstructionCount requestCount =
-        InstructionCount.builder()
-            .actionTypeName(actionType.getName())
-            .instruction(DistributionInfo.Instruction.REQUEST)
-            .count(0)
-            .build();
-    final InstructionCount cancelCount =
-        InstructionCount.builder()
-            .actionTypeName(actionType.getName())
-            .instruction(DistributionInfo.Instruction.CANCEL_REQUEST)
-            .count(0)
-            .build();
-
+  private void processActionType(final ActionType actionType) {
+    String actionTypeName = actionType.getName();
+    RLock lock = redissonClient.getFairLock(LOCK_PREFIX + actionTypeName);
     try {
-      final List<Action> actions = retrieveActions(actionType);
-      processActions(actions, requestCount, cancelCount);
-    } catch (final Exception e) {
-      log.with("action_type", actionType.getName()).error("Failed to process action type", e);
+      if (lock.tryLock(appConfig.getDataGrid().getLockTimeToLiveSeconds(), TimeUnit.SECONDS)) {
+        try {
+          List<Action> actions =
+              actionRepo.findSubmittedOrCancelledByActionTypeName(
+                  actionType.getName(), appConfig.getActionDistribution().getRetrievalMax());
+          actions.forEach(this::processAction);
+        } finally {
+          // Always unlock the distributed lock
+          lock.unlock();
+        }
+      }
+    } catch (InterruptedException ex) {
+      // Ignored - process stopped while waiting for lock
     }
-    return Arrays.asList(requestCount, cancelCount);
   }
 
-  private void processActions(
-      final List<Action> actions,
-      final InstructionCount requestCount,
-      final InstructionCount cancelCount) {
-    log.with(actions).debug("Dealing with actions");
-    for (final Action action : actions) {
-      try {
-        processAction(action, requestCount, cancelCount);
-      } catch (final Exception e) {
-        log.with("action_id", action.getId())
-            .error("Could not process action, will be retried at next scheduled distribution", e);
+  private void processAction(Action action) {
+    ActionProcessingService ap = getActionProcessingService(action);
+
+    // If social reminder action type then generate new IAC
+    if (action.getActionType().getActionTypeNameEnum()
+        == uk.gov.ons.ctp.response.action.representation.ActionType.SOCIALREM) {
+      caseSvcClientService.generateNewIacForCase(action.getCaseId());
+    }
+
+    try {
+      if (action.getState().equals(ActionState.SUBMITTED)) {
+        ap.processActionRequests(action);
+      } else if (action.getState().equals(ActionState.CANCEL_SUBMITTED)) {
+        ap.processActionCancel(action);
       }
+    } catch (Exception ex) {
+      // We intentionally catch all exceptions here.
```

Comment:
perhaps we should put this all the way around the method contents, if it really is an intentional catch-all which would have been one level higher, but can't be because the use of lambda/foreach/`::` convenience stuff.

Reply:
I had a look at doing this but I found that it made the `processActionType` method look a bit overcomplicated. There's probably some more refactoring that could be done around here generally but might be worth saving for another PR